Source/WebKit/UIProcess/WebProcessActivityState.h

```diff
@@ -27,45 +27,55 @@
 
 #include <wtf/WeakRef.h>
 
+namespace WebCore {
+struct PageIdentifierType;
+using PageIdentifier = LegacyNullableObjectIdentifier<PageIdentifierType>;
+}
+
 namespace WebKit {
 
 class ProcessThrottlerActivity;
+class ProcessThrottlerTimedActivity;
 class WebProcessProxy;
 
 class WebProcessActivityState {
     WTF_MAKE_FAST_ALLOCATED;
 public:
     explicit WebProcessActivityState(WebProcessProxy&);
-    void takeVisibleActivity();
-    void takeAudibleActivity();
-    void takeCapturingActivity();
+    void takeVisibleActivity(WebCore::PageIdentifier);
+    void takeAudibleActivity(WebCore::PageIdentifier);
+    void takeCapturingActivity(WebCore::PageIdentifier);
 
-    void reset();
-    void dropVisibleActivity();
-    void dropAudibleActivity();
-    void dropCapturingActivity();
+    void reset(WebCore::PageIdentifier);
+    void dropVisibleActivity(WebCore::PageIdentifier);
+    void dropAudibleActivity(WebCore::PageIdentifier);
+    void dropCapturingActivity(WebCore::PageIdentifier);
 
-    bool hasValidVisibleActivity() const;
-    bool hasValidAudibleActivity() const;
-    bool hasValidCapturingActivity() const;
+    bool hasValidVisibleActivity(WebCore::PageIdentifier) const;
+    bool hasValidAudibleActivity(WebCore::PageIdentifier) const;
+    bool hasValidCapturingActivity(WebCore::PageIdentifier) const;
 
 #if PLATFORM(IOS_FAMILY)
-    void takeOpeningAppLinkActivity();
-    void dropOpeningAppLinkActivity();
-    bool hasValidOpeningAppLinkActivity() const;
+    void takeOpeningAppLinkActivity(WebCore::PageIdentifier);
+    void dropOpeningAppLinkActivity(WebCore::PageIdentifier);
+    bool hasValidOpeningAppLinkActivity(WebCore::PageIdentifier) const;
+#endif
+
+#if PLATFORM(MAC)
+    void takeWasRecentlyVisibleActivity(WebCore::PageIdentifier);
 #endif
 
 private:
     WeakRef<WebProcessProxy> m_process;
 
-    std::unique_ptr<ProcessThrottlerActivity> m_isVisibleActivity;
+    HashMap<WebCore::PageIdentifier, UniqueRef<ProcessThrottlerActivity>> m_isVisibleActivity;
```

Comment:
Maybe they should be stored on the WebPageProxy then? Like they used to?

There is also less chance of leaking assertions that way.

Reply:
Having a WebProcess keeping itself running via process assertions stored in HashMap like this is just bound to leak process assertions and lead to power problems. We have to be super careful to clean those data structures up whenever a page is going away. Whereas, if those are stored on the page, they're guaranteed to go away with the page and not leak, so that an unused process would reliably suspend.

Reply:
There needs to be a way to manage process assertions for iframe processes. Doing that from WebPageProxy seemed weird, but I can probably move it back there.

What if I changed this to be a WeakHashMap with WebPageProxy as the key? They should go away with the page then too.

Reply:
If we need a weak hashmap of pages on the process then we might as well return to the old design, store them on the page (or frame for site isolation) and avoid the hash map altogether.

Process activities were meant to be taken by objects using a process, to keep it running. Having the process taking it to keep itself running doesn’t make as much sense logically and is riskier in terms of process assertion leaks. The page/frame is playing media inside the process and thus keeps the process running due to media playback.